### ISSUE TITLE
Performance issue - High transfer costs

### ISSUE THREAD

**Issue by matthieubulte** [rapidsai/dask-cuda#93]
I am currently working on implementing a simple logistic regression model on a multi-gpu machine with `dask-cuda` and `cupy`, but I'm currently running in some performance issues due to high communication costs (`transfer-*` tasks).

## The setup
I'm running the code on a EC2 `g3.8xlarge` instance using a `LocalCUDA` cluster with the default configuration:
```python
cluster = LocalCUDACluster()
client = Client(cluster)
```
With
- `dask: 2.1.0`
- `dask_cuda: latest 0.9.0`
- `cupy: 6.1.0`

## The data
I use synthetic data, and persist the results (and wait until the cluster is done computing them!) to make sure that my I'm only considering the computation cost of the gradient/loss evaluation:
```python
rs = da.random.RandomState()

X = rs.normal(10, 1, size=(10**7, 99), chunks=(8*10**5, 99))
y = rs.normal(10, 1, size=(10**7,), chunks=(8*10**5,))
beta = rs.normal(10, 1, size=(X.shape[1],))

X = X.map_blocks(cp.array, dtype=np.float64)
y = y.map_blocks(cp.array, dtype=np.float64)
beta = beta.map_blocks(cp.array, dtype=np.float64)

X, y, beta = client.persist([X, y, beta])
```

## The task
I'm currently investigating the cost of a single gradient / loss evaluation. The code for computing it is the following:
```python
p = 1 / (1 + da.exp(-X.dot(beta)))
loss = -da.log(1.0 - y + (2.0*y - 1.0)*p).sum()
grad = (p - y).dot(X)
```

## The problem
Most of the time needed for the computation described above is spent transferring data around, to the point that there is no performance gain between using this 2 GPU instance and a similar machine with 1 GPU (`g3s.xlarge`), and it's slower than using the same machine with a `LocalCluster(n_workers=2)`.

The problem is obvious when looking at the task section of the dashboard:
![tssp](https://user-images.githubusercontent.com/6624306/60965371-f9142b00-a315-11e9-847b-7d15dc28769a.png)

![tsp](https://user-images.githubusercontent.com/6624306/60965356-f0bbf000-a315-11e9-9813-5f2b5d9661fc.png)

![ts](https://user-images.githubusercontent.com/6624306/60965335-e568c480-a315-11e9-8f14-20e8a4b71491.png)

## My question
Is this a `dask_cuda` issue?
Am I doing anything wrong?

More importantly: what would you suggest me to do or investigate to improve this issue?

Thanks!

## Misc
- Here's the optimized graph associated with the above computation:
![graph](https://user-images.githubusercontent.com/6624306/60965480-40022080-a316-11e9-88f3-bc06e01acdba.png)

- Here's the tasks timeline when using 2 workers and 1 thread per worker:
![no_gpu](https://user-images.githubusercontent.com/6624306/60965521-57d9a480-a316-11e9-8ebf-83c0bc961f90.png)

--

**Comment by pentschev** [comment]
This is an issue with Dask using TCP sockets for communication. We are working on UCX to improve that, but I don't know what's the full status of that. I imagine it's only leveraging InfiniBand and NVLink, and it's still very much in development/experimental.

I'm not very familiar with EC2 instances, do you know if they have InfiniBand or NVLink support? If they do, you may be able to run something already, but it's probably not easy. [This post](https://blog.dask.org/2019/06/09/ucx-dgx) UCX setup, but we can't guarantee it's up-to-date nor stable for use yet.

@mrocklin do you know if there's already any work towards shared memory as well and if UCX is usable by the community already?

**Comment by matthieubulte** [comment]
Very interesting, thanks! The instance I'm using (`g3.8xlarge`) does not support NVLink or InfiniBand, but I could use their newest available GPU instances (`p3.x` generation) which do support NVLink.

I will look into the blog post, thank you!

This is maybe more related to Dask itself (or the scheduler), but I'm not sure why so much communication is needed. Implementing this computation by hand could be done with almost no data transfers since every piece of the computation (until `.sum()` is called) can be done locally on each partition, thus locally on the GPU, and the data to transfer is very small (1 gradient + 1 loss per partition = 100 floats per partition).

Are there some debug options to look into this? Also are you aware of any work in the graph optimization or scheduler to reduce data transfers?

**Comment by matthieubulte** [comment]
So, I did some more profiling and found out that the issue is very different than what I expected.

The problem here is that `cupy` computations are (partially at least?) async. This means that doing some stupid computation like:
```
x = cp.ones(10**8)
for i in range(10):
    r += (x+x**2+x**3+x**4).dot(x+x**2+x**3+x**4)
```

Will return very quickly. However, monitoring the GPU shows that it is still running for a while after the python execution has stopped. Notice that in the code presented above, there is no action inducing a sync between GPU and CPU.

Running the same code with a `print(r)` at the end will make the execution much longer and in sync with the GPU activity that can be observed via monitoring.

This means that the `compute` time that `distributed` is collecting here is not reflecting the actual time.

I think this is a bug. I'm not sure which library should be fixing this, but it completely invalidates the performance metrics reported by `distributed`.

I will work on a solution, but I'd love to have your opinion before dong anything PR-worthy.

**Comment by pentschev** [comment]
> Are there some debug options to look into this? Also are you aware of any work in the graph optimization or scheduler to reduce data transfers?

Not sure if this question is still relevant for you, but I'm not very familiar with dask distributed debugging. If it's still relevant, you may have a better chance of getting a good answer if you ask it in stackoverflow or https://github.com/dask/distributed.

> This means that the `compute` time that `distributed` is collecting here is not reflecting the actual time.

Your comment makes a lot of sense, you're most likely onto something. I guess the async behavior of CuPy is not a known case for dask distributed, and that's why it assumes that it's data transfer.

> I think this is a bug. I'm not sure which library should be fixing this, but it completely invalidates the performance metrics reported by distributed.

This is also unclear to me. As a rule-of-thumb, I would say if we're specializing (e.g., for CuPy), it should be here, but if there's a general solution it should be in dask distributed. In the latter case, it would be useful to find an example that is independent of CuPy or any other GPU-dependent libraries to support that and maybe add that as a test.

**Comment by matthieubulte** [comment]
> Not sure if this question is still relevant for you, but I'm not very familiar with dask distributed debugging. If it's still relevant, you may have a better chance of getting a good answer if you ask it in stackoverflow or https://github.com/dask/distributed.

Thanks, yes it makes sense to ask these people.

> I guess the async behavior of CuPy is not a known case for dask distributed, and that's why it assumes that it's data transfer.

Exactly!

> This is also unclear to me. As a rule-of-thumb, I would say if we're specializing (e.g., for CuPy), it should be here, but if there's a general solution it should be in dask distributed. In the latter case, it would be useful to find an example that is independent of CuPy or any other GPU-dependent libraries to support that and maybe add that as a test.

Good point, I will create an issue in the `distributed` repository!
